### PR TITLE
Add `react-native-svg` interface

### PR DESCRIPTION
**Pull request by latekvo**
## Description

This PR adds integration with the `react-native-svg` (`RNSVG`) to improve hitbox detection of the SVG elements they provide.

~~blocked~~ by https://github.com/software-mansion/react-native-svg/pull/2583
~~blocked~~ by `nested SvgViews with viewBox prop have invalid hit detection` - No issue opened yet

## Test plan

- use the newly added SVG integration example for testing this feature

### CODE REVIEW

**Review by m-bert — COMMENTED** [review]
Overall looks ok, left some comments

File: example/App.tsx [hunk 1]
```diff
@@ -40,12 +40,16 @@ import PointerType from './src/release_tests/pointerType';
 import SwipeableReanimation from './src/release_tests/swipeableReanimation';
 import NestedGestureHandlerRootViewWithModal from './src/release_tests/nestedGHRootViewWithModal';
 import TwoFingerPan from './src/release_tests/twoFingerPan';
+import Svg from './src/release_tests/svg';
```
[the remark this answers is not in the document]
Comment (latekvo) [reply example/App.tsx:43]: How about `SvgCompatibility`?

File: android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        view as SvgView
+      }
+
+      while (isSvgElement(highestOrderSvgView.parent)) {
+        if (highestOrderSvgView.parent is SvgView) {
+          highestOrderSvgView = highestOrderSvgView.parent as SvgView
+        }
+        if (highestOrderSvgView.parent is VirtualView) {
+          highestOrderSvgView = (highestOrderSvgView.parent as VirtualView).svgView!!
+        }
```
[the remark this answers is not in the document]
Comment (latekvo) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: Done in [edbe17f](https://github.com/software-mansion/react-native-gesture-handler/pull/3242/commits/edbe17f7a1d1aff7ac84465e0bf7208b0ce1dd78)

File: android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+import android.view.View
+import androidx.core.view.children
+import com.horcrux.svg.SvgView
+import com.horcrux.svg.VirtualView
+
+class RNSVGHitTester {
+  companion object {
+    private fun getHighestOrderSvgView(view: Any): SvgView {
+      var highestOrderSvgView: SvgView
```
[the remark this answers is not in the document]
Comment (latekvo) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: Sure, `getRootSvgView` also works great. Done in [5a767c2](https://github.com/software-mansion/react-native-gesture-handler/pull/3242/commits/5a767c24d93320be20475715d217dfcd2b5e3843)

File: android/nosvg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+class RNSVGHitTester {
+  companion object {
+    @Suppress("UNUSED_PARAMETER")
+    fun isSvgElement(view: Any): Boolean {
+      return false
+    }
+
+    @Suppress("UNUSED_PARAMETER", "COMMENT_IN_SUPPRESSION")
+    fun hitTest(view: Any, posX: Float, posY: Float): Boolean {
+      return false // never accessed
```
[the remark this answers is not in the document]
Comment (latekvo) [reply android/nosvg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt:12]: I think this comment clarifies that this `hitTest` is left to be always `false` intentionally, as it will never be executed.
But I don't have a strong opinion on it, if you'd prefer for this comment to be removed i can do that right away.

>Now it looks like `hitTest` is never called, which I assume is not true if the SVG is present in the project.

It is true, this file will never be loaded, if SVG is present in the project.

File: example/src/release_tests/svg/index.tsx [hunk 5]
```diff
@@ -0,0 +1,117 @@
+import React from 'react';
+import { Text, View, StyleSheet } from 'react-native';
+import { Gesture, GestureDetector } from 'react-native-gesture-handler';
+
+import Svg, { Circle, Rect } from 'react-native-svg';
+
+export default function EmptyExample() {
```
[the remark this answers is not in the document]
Comment (latekvo) [reply example/src/release_tests/svg/index.tsx]: Done in [7660091](https://github.com/software-mansion/react-native-gesture-handler/pull/3242/commits/7660091346d71c30af4d76b660af879c819830a5)

File: android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+        val pressIsInBounds = 0 < posX && posX < view.width && 0 < posY && posY < view.height
+
+        return (hasBeenPressed || hasChildBeenPressed) && pressIsInBounds
+      }
+
+      if (view is VirtualView) {
+        return view.id == highestOrderSvgView.reactTagForTouch(rootX, rootY)
+      }
+      return false
```
[the remark this answers is not in the document]
Comment (latekvo) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: Are you sure we want to do that? 
To me it seems the simplified version, especially the return statement, is much less readable. 
The `if` statement also has a double purpose of triggering the [smart-cast](https://kotlinlang.org/docs/typecasts.html#smart-casts) of `view` from `View` to `SvgView`. 
Without the if statement, I'll have to explicitly cast the `view` to `SvgView`, which will also contribute to making the code less readable.

[the remark this answers is not in the document]
Comment (latekvo) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: >Also it would be nice to use .. instead of <= in bounds checks, but 0 might be a bit problematic.

Thankfully the ranges in Kotlin are inclusive, so `0` won't cause any issues.

Done in [543a8fe](https://github.com/software-mansion/react-native-gesture-handler/pull/3242/commits/543a8fe454c1e7ba6f90494d1f1a71599a8454e6)

[the remark this answers is not in the document]
Comment (latekvo) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: >Speaking of which, is it possible to detect click out of bounds, so that posX will be for example -1?

I think that should be currently possible, the `isWithinBounds` call is executed for every gesture handler no matter it's position on the screen.

File: example/App.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (m-bert) [reply example/App.tsx:43]: Sounds good. I want to avoid using `Svg` as example name as it is mostly used as component's name, so I guess that anything that you'll come up with will be fine 😅

File: android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (m-bert) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: I think that `svgRootView` sounds a bit better, but that is not a strong opinion and we can use `rootSvgView` 😅

File: android/nosvg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (m-bert) [reply android/nosvg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt:12]: > This file isn't loaded when SVG is present in the project.

I know, but it has its counterpart in `android/svg/...`, so basically this function is always present, but its source changes. Thats why this comment sounds a bit confusing.

File: example/App.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (latekvo) [reply example/App.tsx]: Done in [cd1fc6d](https://github.com/software-mansion/react-native-gesture-handler/pull/3242/commits/cd1fc6daf56238f78ea78878df8b991bec6b6e8b)

File: android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (m-bert) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: > To me it seems the simplified version, especially the return statement, is much less readable.

For me it doesn't make much of a difference if we use `pressedId == view.id` instead of `hasBeenPressed`, it is simple enough to understand and I wouldn't call it _**much**_ less readable.

> The if statement also has a double purpose of triggering the [smart-cast](https://kotlinlang.org/docs/typecasts.html#smart-casts) of view from View to SvgView.

Okay, that complicates things a bit. Adding 4 more casts may not be the cleanest solution. I also don't like stuff like `x = x as ...` (at least not in `TypeScript`). As far as I can see `Kotlin` doesn't support union types, so I'm not sure if there's much that we can do (cc kotlin master @j-piasecki).

> Are you asking about this in context of hitSlop?

No, I was just wondering whether we have to check if `posX >= 0`, since coordinates are view-relative.

File: android/nosvg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (latekvo) [reply android/nosvg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: Makes sense, removed the comment in [a2c2ffe](https://github.com/software-mansion/react-native-gesture-handler/pull/3242/commits/a2c2ffec096f6f2bd98a6da5e80b5347608e1890)

File: android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (latekvo) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: > Adding 4 more casts may not be the cleanest solution

Only 1 cast will be necessary, but it's still not ideal.

> No, I was just wondering whether we have to check if posX >= 0, since coordinates are view-relative.

I see. The received coordinates are often negative, so the `posX >= 0` is indeed necessary.

> For me it doesn't make much of a difference if we use pressedId == view.id instead of hasBeenPressed, it is simple enough to understand and I wouldn't call it much less readable.

It reduces the self-descriptiveness of both `pressedId == view.id` and `pressedId in childrenIds`, making it a bit harder to read, but i agree it's not that much of a difference. 
If you prefer the simpler notation a lot, i'll change it right away.

[the remark this answers is not in the document]
Comment (m-bert) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: > Only 1 cast will be necessary, but it's still not ideal.

Do you mean only in first occurrence of `view`? If it's only 1 cast than it sounds ok. 

> If you prefer the simpler notation a lot, i'll change it right away.

It is not a strong opinion, it is just a fact that I don't think it is necessary to add new variables for obvious stuff. If for you it becomes less readable than you can leave it.

[the remark this answers is not in the document]
Comment (latekvo) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: > Do you mean only in first occurrence of view? If it's only 1 cast than it sounds ok.

Yes, only the first occurrence.

But as I said, just looking at `hitTest()` function, it's not in any way clear that `view` is limited to being `VirtualView` and `SvgView` without looking at both the usages of `hitTest()`, and the rest of `RNSVGHitTester`.
Having `if (view is SvgView)` and `if (view is VirtualView)` is the only clear indication within `hitTest()` that `view` can be only either one of those. 
Removing that indication in favour of an early return for `VirtualView`, and then casting the `view` to `SvgView` in the rest of the code may seem arbitrary to someone unfamiliar with the code, so I would prefer to keep it the way it is now.

[the remark this answers is not in the document]
Comment (m-bert) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: > (...) it's not in any way clear that view is limited to being `VirtualView` and `SvgView` (...)

I wouldn't say so since we are in the `RNSVGHitTester` file. Also, one line above `hitTest` there's `isSvgElement` method that also indicates which ones are from `SVG`. 

The thing is, I don't like the idea of using `if` as cast if we can use cast directly (as you said, there's only one that is needed) 

> The if statement also has a double purpose of triggering the [smart-cast](https://kotlinlang.org/docs/typecasts.html#smart-casts) of `view` from `View` to `SvgView`.

Now I'd say that it is its only purpose, am I right?

> Removing that indication in favour of an early return for `VirtualView`, and then casting the view to `SvgView` in the rest of the code may seem arbitrary to someone unfamiliar with the code, so I would prefer to keep it the way it is now.

It is not really about early return, it is because this `if` is unnecessary. Also, even if it would be arbitrary, we can leave a comment explaining that these are the only two possibilities. 

What's your stance on this, @j-piasecki?

[the remark this answers is not in the document]
Comment (j-piasecki) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: > I wouldn't say so since we are in the RNSVGHitTester file.

That assumes that the person reading is familiar with how RNSVH is implemented (I don't).

> Also, one line above hitTest there's isSvgElement method that also indicates which ones are from SVG.

That assumes that the person reading has read the entire file (I didn't).

> Having if (view is SvgView) and if (view is VirtualView) is the only clear indication within hitTest() that view can be only either one of those.

Actually, that's not a clear indication because `return false` suggests that there are more cases that we are ignoring. In an ideal world, we would just assert the view to be of the two types or throw instead of returning, but then we risk apps hard-crashing on the integration layer when SVG adds a new view type (which, I think, is not desirable). IMO, a comment that those two should be the only options would be useful.

> I don't like the idea of using if as cast

But this is the (Kotlin) way, that's why smart casts are a feature (off-top: [TS does a similar thing](https://www.typescriptlang.org/docs/handbook/advanced-types.html))

> this if is unnecessary

That's true in runtime (assuming SVG doesn't change its implementation). But this is also a clear sign (to the person reading) that inside the view has a narrower type. Cast does essentially the same thing, but it's much easier to miss and isn't a hard source of truth (in general, cast can be wrong, and type check can also be wrong, but you must really want to break it for it to be the case).

TL;DR: I think ifs are more readable and more future-proof.

[the remark this answers is not in the document]
Comment (m-bert) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: Okay, you've convinced me

[the remark this answers is not in the document]
Comment (j-piasecki) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: I'd say first comment in the location of the second (closer to where it's applicable)

[the remark this answers is not in the document]
Comment (latekvo) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: Done in [04f46dd](https://github.com/software-mansion/react-native-gesture-handler/pull/3242/commits/04f46dd980c3f17b35ec506c8f454da033163a1e).

[the remark this answers is not in the document]
Comment (latekvo) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: Changes in [a6ba30a](https://github.com/software-mansion/react-native-gesture-handler/pull/3242/commits/a6ba30a888a73a4f0a9eee6b5dac75276604bf2f) made `view is VirtualView` completely redundant, i removed it in [84f5c48](https://github.com/software-mansion/react-native-gesture-handler/pull/3242/commits/84f5c480377e915c055a9c3eebf09fcf92f26462).

File: android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+          (rootSvgView.parent as VirtualView).svgView!!
+        } else {
+          rootSvgView.parent as SvgView
+        }
+      }
+
+      return rootSvgView
+    }
+
+    fun isSvgElement(view: Any): Boolean {
```
Comment (jakex7) [inline android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt:30]: I think we can narrow it down to `View` instead of `Any`

File: android/nosvg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package com.swmansion.gesturehandler
+
+class RNSVGHitTester {
+  companion object {
+    @Suppress("UNUSED_PARAMETER")
+    fun isSvgElement(view: Any) = false
+
+    @Suppress("UNUSED_PARAMETER")
+    fun hitTest(view: Any, posX: Float, posY: Float) = false
```
Comment (jakex7) [inline android/nosvg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt:9]: `nosvg` functions signatures should probably match the `svg/[...]/RNSVGHitTester.kt`

File: android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt (hunk 7 above)
Comment (latekvo) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt:30]: There are cases where `view` is `ViewParent`, which doesn't inherit directly from `View`.
I can look into it further if you'd want, but for now `View` doesn't work there.

File: android/nosvg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt (hunk 8 above)
Comment (latekvo) [reply android/nosvg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt]: Fixed in [29e8492](https://github.com/software-mansion/react-native-gesture-handler/pull/3242/commits/29e8492ef639f364e6b49692fecdcd3f33719f99)

File: android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt (hunk 7 above)
Comment (jakex7) [reply android/svg/src/main/java/com/swmansion/gesturehandler/RNSVGHitTester.kt:30]: Right, I've missed the second call where it can be `ViewParent`

**Review by m-bert — APPROVED** [review]

File: android/build.gradle [hunk 9]
```diff
@@ -45,9 +45,9 @@ def resolveReactNativeDirectory() {
     }
 
     throw new Exception(
-        "[react-native-gesture-handler] Unable to resolve react-native location in " +
-            "node_modules. You should add project extension property (in app/build.gradle) " +
-            "`REACT_NATIVE_NODE_MODULES_DIR` with path to react-native."
+            "[react-native-gesture-handler] Unable to resolve react-native location in " +
+                    "node_modules. You should add project extension property (in app/build.gradle) " +
+                    "`REACT_NATIVE_NODE_MODULES_DIR` with path to react-native."
```
Comment (m-bert) [inline android/build.gradle:50]: Is this done by a formatter?

File: android/build.gradle [hunk 10]
```diff
@@ -75,6 +75,22 @@ def shouldUseCommonInterfaceFromReanimated() {
...
+        return false
+    }
+
+    def inputFile = new File(rnsvg.projectDir, '../package.json')
+    def json = new JsonSlurper().parseText(inputFile.text)
+    def rnsvgVersion = json.version as String
+    def (major, minor, patch) = rnsvgVersion.tokenize('.')
+    return (Integer.parseInt(major) == 15 && Integer.parseInt(minor) == 11 && Integer.parseInt(patch) >= 2) ||
+            (Integer.parseInt(major) == 15 && Integer.parseInt(minor) > 11) ||
+            Integer.parseInt(major) > 15
```
Comment (m-bert) [inline android/build.gradle:91]: Comment above states that it is available from 15.11.0, but here we check for 15.11.2

**Review by m-bert — CHANGES_REQUESTED** [review]

File: android/build.gradle (hunk 9 above)
Comment (latekvo) [reply android/build.gradle:50]: Yes, the formatter.
I debated myself about removing these changes, but figured there's no benefit for deferring these changes.

File: android/build.gradle (hunk 10 above)
Comment (latekvo) [reply android/build.gradle:91]: Fixed the comment in [98731e5](https://github.com/software-mansion/react-native-gesture-handler/pull/3242/commits/98731e5c305e61142b6e7335c30d42b960f21edb). The correct version is `15.11.2`.

File: android/build.gradle [hunk 11]
```diff
@@ -211,15 +233,20 @@ def kotlin_version = safeExtGet('kotlinVersion', project.properties['RNGH_kotlin
...
-    
+
 
     if (shouldUseCommonInterfaceFromReanimated()) {
         // Include Reanimated as dependency to load the common interface
-        implementation (rootProject.subprojects.find { it.name == 'react-native-reanimated' }) {
-            exclude group:'com.facebook.fbjni' // resolves "Duplicate class com.facebook.jni.CppException"
+        implementation(rootProject.subprojects.find { it.name == 'react-native-reanimated' }) {
+            exclude group: 'com.facebook.fbjni'
+            // resolves "Duplicate class com.facebook.jni.CppException"
```
Comment (m-bert) [inline android/build.gradle:242]: If this is also done by formatter, could you please move this comment line above?

Also, what formatter do you use? I do not recall that our formatting script changes `gradle` files.

Comment (latekvo) [reply android/build.gradle]: moved comment in [423ac1c](https://github.com/software-mansion/react-native-gesture-handler/pull/3242/commits/423ac1c0ff0ee295ae35ea6f9b5a4bb3ed3798f4)

**Review by m-bert — APPROVED** [review]
